Documentation for the Authorization filter (#2992)
kroxylicious/kroxylicious: kroxylicious-docs/docs/_modules/authorization/con-authorization-rules.adoc

:_mod-docs-content-type: CONCEPT

[id='con-authorization-rules-{context}']
= Authorization rules

The authorization rules define which principals can perform specific operations on specific Kafka resources.

== Outline of a rule file

The following example shows the overall outline of a rule file.
The sections that follow give more details.

[source]
----
// Comment
from io.kroxylicious.filter.authorization import TopicResource as Topic;

deny User with name = "alice" to * Topic with name = "payments-received";
allow User with name = "alice" to * Topic with name like "payments-*";

otherwise deny;
----

== Comments

> Both line and block comments are supported.  Line comments are proceeded by `//`.

Reviewer: ```suggestion
Both line and block comments are supported.  Line comments are preceded by `//`.
```

To declare it use a `import` statement like this.

[source]
----
from io.kroxylicious.filter.authorization import TopicResource;
----

To declare it with an alias, use an `import` statement like this:

[source]
----
from io.kroxylicious.filter.authorization import TopicResource as Topic;
----

== Rules

The basic form of a rule is as follows:

[source]
----
<allow|deny> User with <user predicate> to <operation> <resource type> with <resource predicate>;
----
where:

* `<allow|deny>` indicates whether to allow or deny the action.
* `<user predicate>` matches the user principal performing the action.
* `<resource type>` identifies the resource type being acted upon. This can either be the name of the resource type name or an alias for it.
* `<resource predicate>` identifies the resource.
* `<operation>` identifies the operation(s) to be performed on the resource.

Rules must be ordered so that any `deny` rules proceed the `allow` rules.

When rules are evaluated, they are considered from top to bottom, with the first matching rule taking precedence.

== Otherwise deny

Rules files must end with the statement `otherwise deny`.

> This stipulation means that all rules files are have deny-by-default semantics.

Reviewer: IE once we support Group authorization, a rule file that does not import Group would be permissive of group operations, so that there is not a change of behaviour when the user deploys the new Filter with their existing Topic-only rules

:_mod-docs-content-type: CONCEPT

[id='con-authorization-rules-{context}']
= Authorization Rules

> The authorization rules define who is allowed to do what.

Reviewer 2: ```suggestion
The authorization rules define which users can perform specific operations on specific Kafka resources.
```

> The authorization rules define who is allowed to do what.

Reviewer: I don't know how to convey this neatly, but it feels a bit clunky launching into rule/import syntax without some higher level description saying something about how we use a custom language for defining ACL rule and the user will write their rules in this form.

Maybe we could instead lead this section with an example and break down the bits of syntax? Having the ACL language examples away from `Authorization Rules` feels a little disconnected.

Reviewer: it makes sense to have the configuration example, but maybe we should additionally have ACL rules examples

The basic form of a rule is as follows:

[source]
----
<allow|deny> User with <user predicate> to <operation> <resource type> with <resource predicate>;
----
where:

* `<allow|deny>` indicates whether to allow or deny the action.
* `<user predicate>` matches the user performing the action.
* `<resource type>` and `<resource predicate>` identifies the resource.
* `<operation>` identifies the operation(s) to be performed on the resource.

> `<resource type>` before be declared before use.

Reviewer 2: ```suggestion
`<resource type>` must be declared before use.
```

This is done with a `from` statement in the following form:

[source]
----
from <package> import <element> [as <alias>][,... , <elementn> [as <aliasn>]];
----
where:

* <package> is `io.kroxylicious.filter.authorization`
* <element> is the `ResourceType` implementation `TopicResource`.

> * <alias> is an optional alias for the resource type.

Reviewer 2: ```suggestion
* <alias> is an optional alias for the resource type.

For example, `TopicResource` is the implementation that represents Kafka topics:

[source]
----
from io.kroxylicious.filter.authorization import TopicResource;
----
```

> Rules files must end with the statement `otherwise deny`.

Reviewer 2: what happens if rules overlap?

Reviewer: was just coming to say this, this high level description of how rules are applied should probably come before we go into the individual rule syntax (and cover how deny rules must precede allow rules)

Author: I'm now describing rule ordering, and how rules are considered top-to-bottom with the first matching rule taking precedence.

[source]
----
...
otherwise deny;
----

> == User Predicates

Reviewer: yep feels like something missing, the glossary defines `Principal` but nothing else mentions it.

The `User` referred to in the rules language is a Principal, contained within the Authenticated Subject.

The following user predicates are supported:

|===
|Predicate|Description|Example

| `=`

> | Equality

Reviewer 2: ```suggestion
| Equals
```

| `name = "alice"`

| `IN`
| Set inclusion
| `name in +{"alice", "bob"}+`

| `LIKE`
| Prefix (note that the wildcard * is only permitted at the end of the prefix.)
| `name like "bob*"`

|===

== Resource Predicates

The following resource predicates are supported:

|===
|Predicate|Description|Example

| `=`
| Equality
| `name = "mytopic"`

| `IN`
| Set inclusion
| `name in +{"topic1", "topic2"}+`

| `LIKE`

> | Prefix (note that the wildcard * is only permitted at the end of the prefix.)

Reviewer 2: ```suggestion
| Prefix (wildcard * is permitted only at the end)
```

| `name like "finance*"`

| `MATCHING`
| Regular expression match
| `name matching "a+"`

|===

== Operations

Rules may specify operations may be specified in the following ways:

* singularly e.g `READ`
* a set of operations e.g. `+{READ, WRITE}+`

> * wildcard e.g. `*`

Reviewer 2: ```suggestion
== Operations

Operations in rules can be specified in the following ways:

* As a single operation, for example `READ`
* As a set of operations, for example `+{READ, WRITE}+`
* As a wildcard that matches any operation, for example `*`
```
